File: basecode/driver_app/lib/features/auth/repository/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/material.dart';
import 'package:logger/logger.dart';

class AuthService {
  final FirebaseAuth _auth = FirebaseAuth.instance;
  Logger logger = Logger();

  // Login with email and password
  Future<User?> loginWithEmailAndPassword(String email, String password) async {
    try {
      logger.f("Singing In");
      UserCredential userCredential = await _auth.signInWithEmailAndPassword(
          email: email, password: password);
      return userCredential.user;
    } catch (e) {
      logger.e('Error logging in: $e');
      return null;
    }
  }

// Send email verification
  Future<void> sendVerificationEmail(User user) async {
    if (!user.emailVerified) {
      try {
        await user.sendEmailVerification();
        logger.i('Verification email sent');
      } catch (e) {
        logger.e('Failed to send verification email: $e');
      }
    }
  }

//Send password recovery email
  Future<void> sendPasswordRecoveryEmail(
      String email, BuildContext context) async {
    try {
      await FirebaseAuth.instance.sendPasswordResetEmail(email: email);
      if (context.mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
              content: Text('Email enviado, revisa tu bandeja de entrada')),
        );
      }

      logger.i('Password recovery email sent');
    } catch (e) {
      if (context.mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
              content: Text(
                  'Error: No se pudo enviar el email para recuperar la contraseña')),
        );
      }
      logger.e('Failed to send password recovery email: $e');
    }
  }

  // Sign out
  Future<void> signOut() async {
    await _auth.signOut();
  }
}

File: basecode/admin/src/app/modules/auth/services/auth.service.ts
import { Injectable } from '@angular/core';
import { AngularFireAuth } from '@angular/fire/compat/auth';
import { MatSnackBar } from '@angular/material/snack-bar';
import { promises } from 'dns';
import { FirebaseErrorCodeService } from '../../../shared/services/FirebaseErrorCode/firebase-error-code.service';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  constructor(
    private afAuth: AngularFireAuth,
    private snackBar: MatSnackBar,
    private firebaseErrorCode: FirebaseErrorCodeService,

    
  ) {}

  //Sign in with Email and Password
  async signIn(email: string, password: string): Promise<any> {
    
    try {
      const userCredential = await this.afAuth.signInWithEmailAndPassword(
        email,
        password
      );
      return userCredential;
    } catch (error: any) {
      console.log('An error occurred while logging into Firebase', error);
      console.log(error);
      this.snackBar.open(
        this.firebaseErrorCode.firebaseError(error.code),
        'Cerrar',
        { duration: 5000 }
      );
      return null;
    }
  }

  //Create driver account with Email and Password
  async createDriverAccount(email: string, password: string): Promise<any> {
    try {
      const userCredential = await this.afAuth.createUserWithEmailAndPassword(
        email,
        password
      );
      return userCredential;
    } catch (error: any) {
      this.snackBar.open(
        this.firebaseErrorCode.firebaseError(error.code),
        'Cerrar',
        { duration: 5000 }
      );
      return null;
    }
  }

   //For user be able to reset his password
   async sendPasswordReset(email: string): Promise<void> {
    try {
      await this.afAuth.sendPasswordResetEmail(email);
      this.snackBar.open('Se ha enviado un enlace para restablecer la contraseña al conductor.', 'Success');
    } catch (error) {
      this.snackBar.open('No se pudo enviar el enlace de restablecimiento de contraseña.', 'Error');
    }
  }

   //Paswword recovery
   async recover(email:string):Promise<void>{
     try {
      await this.afAuth.sendPasswordResetEmail(email);
      this.snackBar.open('Te hemos enviado un correo para reestablecer tu contraseña','Cerrar');
     } catch (error:any) {
      this.snackBar.open(this.firebaseErrorCode.firebaseError(error.code),'Cerrar');
     }
  }

  //Sign out
  signOut(): Promise<any> {
    return this.afAuth.signOut();
  }
}
